File: src/part01/chapter02/levelC/task01/java/util/TimesValueInArrayCounter.java
package part01.chapter02.levelC.task01.java.util;

import java.util.Arrays;

public class TimesValueInArrayCounter {

    public static double[][] obtainTimesInArrayValueAndCount(double[] array) {
        double[][] result = new double[0][2];
        for (double number : array) {
            result = addTimesToTimesArray(result, number);
        }
        return result;
    }

    public static int getTimes(double[][] timesArray, double number) {
        if (containsInTimesArray(timesArray, number)) {
            return (int) timesArray[indexOfTimesArray(timesArray, number)][1];
        }
        return 0;
    }

    public static double[][] removeTimeFromTimesArray(double[][] timesArray, double number) {
        double[][] result = timesArray.clone();
        if (containsInTimesArray(timesArray, number)) {
            result[indexOfTimesArray(result, number)][1]--;
        }
        return result;
    }

    private static double[][] addTimesToTimesArray(double[][] timesArray, double number) {
        double[][] result;
        if (containsInTimesArray(timesArray, number)) {
            result = timesArray.clone();
            result[indexOfTimesArray(result, number)][1]++;
        } else {
            result = Arrays.copyOf(timesArray, timesArray.length + 1);
            result[result.length - 1] = new double[]{number, 1.0};
        }
        return result;
    }

    private static boolean containsInTimesArray(double[][] timesArray, double number) {
        return indexOfTimesArray(timesArray, number) != -1;
    }

    private static int indexOfTimesArray(double[][] timesArray, double value) {
        for (int i = 0; i < timesArray.length; i++) {
            if (timesArray[i][0] == value) {
                return i;
            }
        }
        return -1;
    }
}

File: src/part01/chapter02/levelC/task01/kotlin/util/TimesValueInArrayCounter.kt
package part01.chapter02.levelC.task01.kotlin.util

import part01.chapter02.levelC.matrix.kotlin.cloneMatrix

object TimesValueInArrayCounter {

    fun obtainTimesInArrayValueAndCount(array: DoubleArray): Array<DoubleArray> {
        var timesArray = arrayOf(DoubleArray(2))
        array.forEach {
            timesArray = addTimeToTimesArray(timesArray, it)
        }
        return timesArray
    }

    fun getTimes(timesArray: Array<DoubleArray>, number: Double): Int {
        if (containsInTimesArray(timesArray, number)) {
            return timesArray[indexOfTimesArray(timesArray, number)][1].toInt()
        }
        return 0
    }

    fun removeTimeFromTimesArray(timesArray: Array<DoubleArray>, number: Double): Array<DoubleArray> {
        val result = cloneMatrix(timesArray)
        if (containsInTimesArray(result, number)) {
            result[indexOfTimesArray(result, number)][1]--
        }
        return result
    }

    private fun addTimeToTimesArray(timesArray: Array<DoubleArray>, number: Double): Array<DoubleArray> {
        return if (containsInTimesArray(timesArray, number)) {
            val result = cloneMatrix(timesArray)
            result[indexOfTimesArray(timesArray, number)][1]++
            result
        } else {
            val result = Array(timesArray.size + 1) { DoubleArray(2) }
            for (rowIndex in timesArray.indices) {
                result[rowIndex] = timesArray[rowIndex]
            }
            result[result.lastIndex] = doubleArrayOf(number, 1.0)
            result
        }
    }

    private fun containsInTimesArray(timesArray: Array<DoubleArray>, number: Double) = indexOfTimesArray(timesArray, number) != -1

    private fun indexOfTimesArray(timesArray: Array<DoubleArray>, number: Double): Int {
        for (rowIndex in timesArray.indices) {
            if (timesArray[rowIndex][0] == number) {
                return rowIndex
            }
        }
        return -1
    }
}